ow_dynamic_terrain/src/TerrainModifier.cpp

```diff
@@ -215,13 +257,35 @@ void TerrainModifier::applyImageToHeightmap(Heightmap* heightmap, const Point2i&
   auto right = min(center.x - image.cols / 2 + image.cols - 1, heightmap_size);
   auto bottom = min(center.y - image.rows / 2 + image.rows - 1, heightmap_size);
 
+  auto diff = TerrainBrush::createZerosMatLike(image);
+
+  bool change_occurred = false;
+
   for (auto y = top; y <= bottom; ++y)
     for (auto x = left; x <= right; ++x)
     {
       auto pixel_value = image.at<float>(y - top, x - left);
       if (skip_zeros && ignition::math::equal<float>(pixel_value, 0.0f))
         continue;
-      auto new_height = merge_method(get_height_value(x, y), pixel_value + z_bias);
+      
+      auto old_height = get_height_value(x, y);
+      auto new_height = merge_method(old_height, pixel_value + z_bias);
+
+      if (old_height == new_height) 
+        continue; // no change is necessary
+      
       set_height_value(x, y, new_height);
+      diff.at<float>(y - top, x - left) = new_height - old_height;
+      
+      // if we make it here, flag that a change has occurred
+      change_occurred = true;
     }
+
+    if (change_occurred)
+    {
+      out_diff_image.image    = diff;
+      out_diff_image.encoding = image_encodings::TYPE_32FC1;
+    }
```

Comment:
Why do you need to check for _change_occured_ before copying the diff? Is it for performance?